### PR TITLE
fix: api: Length check the array sent to RPC.

### PR DESCRIPTION
**Pull request by parthshah1**
## Related Issues

When calling **eth_feeHistory**, it takes input a rewardPercentile array. Currently, there isn't any restriction on the length of the array, can result in high computation and potentially leading to OOM issues in RPC nodes. The fix is to set a hard limit of the array to 100. 
## Proposed Changes

Changing the length of RewardPercentiles Array to upto 100.
## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] If the PR affects users (e.g., new feature, bug fix, system requirements change), update the CHANGELOG.md and add details to the UNRELEASED section.
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: node/impl/full/eth.go [hunk 1]
```diff
@@ -687,6 +687,9 @@ func (a *EthModule) EthFeeHistory(ctx context.Context, p jsonrpc.RawParams) (eth
 	if params.BlkCount > 1024 {
 		return ethtypes.EthFeeHistory{}, fmt.Errorf("block count should be smaller than 1024")
 	}
+	if len(*params.RewardPercentiles) >= 100 {
+		return ethtypes.EthFeeHistory{}, fmt.Errorf("Length cannot be greater than 100")
+	}
 	rewardPercentiles := make([]float64, 0)
 	if params.RewardPercentiles != nil {
```
Comment (rvagg) [inline node/impl/full/eth.go:694]:
```suggestion
	rewardPercentiles := make([]float64, 0)
	if params.RewardPercentiles != nil {
		if len(*params.RewardPercentiles) >= 100 {
			return ethtypes.EthFeeHistory{}, fmt.Errorf("Length cannot be greater than 100")
		}
```

@parthshah1 you've introduced a panic DoS in your DoS fix!

**Comment by rvagg** [comment]
Seems reasonable, except that I don't know what a reasonable limit on this would be; geth doesn't appear to have one, at least when you get this deep: https://github.com/ethereum/go-ethereum/blob/c41105ce80f12f60ec4bf6c65c4c59c6bf4a86e7/eth/gasprice/feehistory.go#L200 and it event does an `8*len` allocation at one point. Maybe there's a total params limit size, which I guess would be appropriate for us too if we don't have one.

File: node/impl/full/eth.go (hunk 1 above)
Comment (parthshah1) [reply node/impl/full/eth.go]: I had a nudge. Just saw CI logs. Apologies.

**Comment by parthshah1** [comment]
@rvagg. I had a discussion about the limits with @Stebalien, that maybe 100 would be reasonable number since one would not need more than that to calculate the eth_feehistory. It also aims to solve computation issues this might cause with huge arrays.  

> it event does an 8*len allocation at one point. Maybe there's a total params limit size, which I guess would be appropriate for us too if we don't have one.

I was not aware of such a case in geth. I am not too expert in this area and how often, how much data it is expected to compute. I will agree to what you think should be right. Though, imo keeping the hard fix on the length won't harm.

**Review by rvagg — APPROVED** [review]
this seems fine to me, we can bump it up if there's a genuine request for more

File: node/impl/full/eth.go [hunk 2]
```diff
@@ -689,6 +689,9 @@ func (a *EthModule) EthFeeHistory(ctx context.Context, p jsonrpc.RawParams) (eth
 	}
 	rewardPercentiles := make([]float64, 0)
 	if params.RewardPercentiles != nil {
+		if len(*params.RewardPercentiles) >= 100 {
```
Comment (Stebalien) [inline node/impl/full/eth.go:692]:
```suggestion
		if len(*params.RewardPercentiles) > 100 {
```

(to match the error description)

Comment (Stebalien) [reply node/impl/full/eth.go:692]: I'd also make the limit a constant at the top of the file.

**Review by Stebalien — COMMENTED** [review]
Nits, but this looks correct.

File: node/impl/full/eth.go (hunk 2 above)
Comment (parthshah1) [reply node/impl/full/eth.go]: Done.

File: node/impl/full/eth.go [hunk 3]
```diff
@@ -40,6 +40,7 @@ import (
 )
 
 var ErrUnsupported = errors.New("unsupported method")
+var RewardPercentileArrayLimit int = 100
```
Comment (Stebalien) [inline node/impl/full/eth.go:43]: nit `maxEthFeeHistoryRewardPercentiles` is a little more descriptive (not a limit on the array size, a limit on something we're doing).

Also, `const ... = 100`, no need for this to be variable.

**Review by Stebalien — APPROVED** [review]

File: node/impl/full/eth.go (hunk 3 above)
Comment (parthshah1) [reply node/impl/full/eth.go]: Made appropriate changes.

**Review by Stebalien — APPROVED** [review]

**Comment by Stebalien** [comment]
Thank you! Sorry for all the nits.

One final thing: it looks like some of the lints are unhappy. I _think_ you just need to run `go fmt` and fix some spelling.

**Review by rvagg — APPROVED** [review]

File: node/impl/full/eth.go [hunk 4]
```diff
@@ -689,6 +691,9 @@ func (a *EthModule) EthFeeHistory(ctx context.Context, p jsonrpc.RawParams) (eth
 	}
 	rewardPercentiles := make([]float64, 0)
 	if params.RewardPercentiles != nil {
+		if len(*params.RewardPercentiles) > maxEthFeeHistoryRewardPercentiles {
+			return ethtypes.EthFeeHistory{}, fmt.Errorf("Length of the reward percentile array cannot be greater than 100")
```
Comment (rvagg) [inline node/impl/full/eth.go:695]:
```suggestion
			return ethtypes.EthFeeHistory{}, errors.New("length of the reward percentile array cannot be greater than 100")
```

**Pull request merged**